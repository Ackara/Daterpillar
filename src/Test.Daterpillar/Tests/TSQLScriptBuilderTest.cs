using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using Gigobyte.Daterpillar.TextTransformation;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Tests.Daterpillar.Constants;
using Tests.Daterpillar.Helpers;

namespace Test.Daterpillar.Tests
{
    [TestClass]
    [DeploymentItem(SampleData.Folder)]
    [DeploymentItem(KnownFile.DbConfig)]
    [UseApprovalSubdirectory(nameof(ApprovalTests))]
    [UseReporter(typeof(FileLauncherReporter), typeof(ClipboardReporter))]
    public class TSQLScriptBuilderTest : DbTemplateBuilderTestBase
    {
        [ClassCleanup]
        public static void Cleanup()
        {
            ApprovalTests.Maintenance.ApprovalMaintenance.CleanUpAbandonedFiles();
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_builds_a_new_schema_when_settings_are_enabled()
        {
            var settings = new ScriptBuilderSettings()
            {
                AppendScripts = true,
                AppendComments = true,
                TruncateDatabaseIfItExist = true
            };

            RunSchemaTest<TSQLScriptBuilder>(settings, DatabaseHelper.CreateMSSQLConnection());
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_builds_a_new_schema_when_settings_are_disabled()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                var settings = new ScriptBuilderSettings()
                {
                    AppendScripts = false,
                    AppendComments = false,
                    TruncateDatabaseIfItExist = false
                };

                RunSchemaTest<TSQLScriptBuilder>(settings, connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_adds_a_new_index_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunIndexTest<TSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_adds_a_new_foreign_key_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunForeignKeyTest<TSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_removes_a_schema_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunSchemaDropTest<TSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_removes_a_table_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunTableDropTest<TSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_that_removes_a_index_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunIndexDropTest<TSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_removes_a_foreign_key_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunForeignKeyDropTest<TSQLScriptBuilder>(connection);
            }
        }

        //[TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Alter_should_generate_a_tsql_modify_script_for_a_table_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunAlterTableTest<TSQLScriptBuilder>(connection);
            }
        }

        //[TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Alter_should_generate_a_tsql_modify_script_for_a_column_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunAlterColumnTest<TSQLScriptBuilder>(connection);
            }
        }
    }
}