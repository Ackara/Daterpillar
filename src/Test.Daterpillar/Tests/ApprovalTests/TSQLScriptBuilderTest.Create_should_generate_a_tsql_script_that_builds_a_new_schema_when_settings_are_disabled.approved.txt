-- ======================================================================
-- NAME:
-- daterpillar1

-- DESCRIPTION:
-- This is a usefull description.

-- AUTHOR:
-- john@example.com

-- DATE:
-- Mon 01, 0001 12:00 AM
-- ======================================================================

IF DB_ID('daterpillar1') IS NULL CREATE DATABASE [daterpillar1];
USE [daterpillar1];

IF OBJECT_ID('daterpillar1.dbo.genre') IS NULL CREATE TABLE [daterpillar1].[dbo].[genre]
(
	[Id] INT NOT NULL PRIMARY KEY IDENTITY(1, 1),
	[Name] VARCHAR(64) NOT NULL
);


IF OBJECT_ID('daterpillar1.dbo.artist') IS NULL CREATE TABLE [daterpillar1].[dbo].[artist]
(
	[Id] INT NOT NULL PRIMARY KEY IDENTITY(1, 1),
	[Name] VARCHAR(64) NOT NULL,
	[Bio] TEXT
);

IF NOT EXISTS(SELECT * FROM [sys].[indexes] WHERE [object_id]=OBJECT_ID('daterpillar1.dbo.artist') AND [name]='artist_idx1') CREATE UNIQUE INDEX [artist_idx1] ON [daterpillar1].[dbo].[artist] ([Name] DESC);

IF OBJECT_ID('daterpillar1.dbo.album') IS NULL CREATE TABLE [daterpillar1].[dbo].[album]
(
	[Artist_Id] INT NOT NULL,
	[Name] VARCHAR(64) NOT NULL,
	[Release_Date] DATE,
	CONSTRAINT [album_artist_id_to_artist_id_fkey2] FOREIGN KEY ([Artist_Id]) REFERENCES [artist]([Id]) ON UPDATE CASCADE ON DELETE CASCADE
);

IF NOT EXISTS(SELECT * FROM [sys].[indexes] WHERE [object_id]=OBJECT_ID('daterpillar1.dbo.album') AND [name]='album_idx3') CREATE INDEX [album_idx3] ON [daterpillar1].[dbo].[album] ([Artist_Id] ASC, [Name] ASC);

IF OBJECT_ID('daterpillar1.dbo.song') IS NULL CREATE TABLE [daterpillar1].[dbo].[song]
(
	[Id] INT NOT NULL PRIMARY KEY IDENTITY(1, 1),
	[Album_Id] INT NOT NULL,
	[Artist_Id] INT NOT NULL,
	[Genre_Id] INT NOT NULL,
	[Name] VARCHAR(256) NOT NULL,
	[Length] DECIMAL(4, 2),
	[Price] DECIMAL(12, 2) NOT NULL,
	[On_Device] BIT NOT NULL,
	CONSTRAINT [song_genre_id_to_genre_id_fkey4] FOREIGN KEY ([Genre_Id]) REFERENCES [genre]([Id]) ON UPDATE CASCADE ON DELETE CASCADE,
	CONSTRAINT [song_artist_id_to_artist_id_fkey5] FOREIGN KEY ([Artist_Id]) REFERENCES [artist]([Id]) ON UPDATE CASCADE ON DELETE CASCADE
);

IF NOT EXISTS(SELECT * FROM [sys].[indexes] WHERE [object_id]=OBJECT_ID('daterpillar1.dbo.song') AND [name]='song_idx') CREATE INDEX [song_idx] ON [daterpillar1].[dbo].[song] ([Name] ASC);

